logger/drain/simple/drain.go

```diff
@@ -0,0 +1,117 @@
+package simple
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"net/url"
+	"sync"
+	"time"
+)
+
+// This determines how often we'll replace an existing connection with a new one.  This ensures we
+// redial, which in turn will ensure any DNS changes affecting where logs are drained to don't
+// go unnoticed for too long.
+const connRefreshInterval = 5 * time.Minute
+
+// This determines how many failed dial attempts are required before the drain is muted.
+const maxFailedConns = 5
+
+// This determines how much time we're willing to spend dialing.
+const dialTimeout = 10 * time.Second
+
+// This is how long the drain is muted for after repeated connection failures.
+const mutePeriod = 5 * time.Minute
+
+type logDrain struct {
+	proto    string
+	uri      string
+	conn     net.Conn
+	connTime time.Time
+	muted    bool
+	mutex    sync.Mutex
+}
+
+// NewDrain returns a pointer to a new instance of a drain.LogDrain
+func NewDrain(drainURL string) (*logDrain, error) {
+	u, err := url.Parse(drainURL)
+	if err != nil {
+		return nil, err
+	}
+	var proto string
+	if u.Scheme == "udp" || u.Scheme == "syslog" {
+		proto = "udp"
+	} else if u.Scheme == "tcp" {
+		proto = "tcp"
+	} else {
+		return nil, fmt.Errorf("Invalid drain url scheme: %s", u.Scheme)
+	}
+	return &logDrain{proto: proto, uri: u.Host + u.Path}, nil
+}
+
+// Send forwards the provided log message to an external destination
+func (d *logDrain) Send(message string) error {
+	if d.muted {
+		return nil
+	}
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	conn, err := d.getConnection(false)
+	if err != nil {
+		return err
+	}
+	_, err = fmt.Fprintln(conn, message)
+	if err != nil {
```

Comment:
I think this is another thing to come back to if there's a demonstrated need for it.

Currently, there are two things that might fail and thus be retried.  A failed message write is probably on account of a broken pipe.  It's retried with a new connection.  Assuming the new connection was successfully obtained, there's no reason the second attempt shouldn't succeed.  So on that front, I'm not sure if we need to get any fancier... at least for now.

The other thing that gets retried are failed dialing attempts.  That's currently hardcoded through a constant to be retried five times before giving up, at which point there isn't a prayer of firing off a log message to the external service, so the drain is temporarily muted.

Muting a log drain has the benefit of (forgive the overloaded term here) _draining_ the internal message queue very quickly so that when things recover, we're draining recent log messages out to the external log archive instead of working through a backlog of older log messages while newer log messages get dropped because of the backed up queue.

I _think_ this all sounds pretty reasonable, but I am more than willing to hear further challenges to that.